### ISSUE TITLE
Documentation we could usefully have

### ISSUE THREAD

**Issue by cannam** [piper-audio/ugly-duckling#46]
Some documentation we could do with.

* Answering the question: If I were to add a new sort of display shape (to go along with the notes, grid, tracks etc shapes), how would I do that? What would I have to add to within each layer of code, from waves-ui up to ugly-duckling? (I ask this because I think it would be good to document the connections between layers in this way, not necessarily because I think people will want to add lots of new shapes)
* Answering the question: How do I write a feature extractor in Javascript for use here?

**Comment by LucasThompson** [comment]
The "display types" and methods for deducing and reshaping to them are spread out across a few different type definitions and functions operating on them in `FeatureUtilities`.

Types:

- `HigherLevelFeatureShape` is a union of different strings / tags, used to identify the deduced shape, currently `'vector' | 'matrix ' | 'tracks' | 'instants' | 'notes'`
- `ShapedFeatureData` is a union of the different data types for the corresponding data to the above
- There are then a bunch of types, `Vector`, `Matrix`, `Tracks`, `Notes`, `Instants`, which pair a tag and a type from above together, using a `shape` and a `collected` property
- There is a generic class / type, `ShapedFeature` which has constraints on using `HigherLevelFeatureShape` and `ShapedFeatureData` which can be used to write these "pairing" types, but they could also just be written by hand - the idea is to get some help from the compiler 
- `KnownShapedFeature` is a union type of the above pairs, affording some help from the compiler when developing. i.e. allowing to switch on the shape property of a `KnownShapedFeature` and knowing the data type of the feature in the case 

It is worth mention that all of the above steps are as such just because that is what is required (or the only way I know of) to end up with something like `KnownShapedFeature`. 

Functions:

- `deduceHigherLevelFeatureShape`, `SimpleResponse -> HigherLevelFeatureShape` - by inspecting the returned features, and / or the output descriptor, decide on a `HigherLevelFeatureShape`
- `toKnownShape`, `SimpleResponse -> KnownShapedFeature`, uses `deduceHigherLevelFeatureShape` internally - from a known display shape, reshape the data in `SimpleResponse` to the corresponding `ShapedFeatureData` and pair with the deduced `HigherLevelFeatureShape`

A call to `FeatureExtractionService::extract` returns a promise which resolves to an `ExtractionResult`, a type which contains a `KnownShapedFeature` in its `result` property. 

In the app root, `AppComponent`, the state of the application is currently represented in a `PersistentStack` of `Item`s. When an extraction request resolves, an `ExtractedAnalysisItem`, which extends `Item`, is pushed onto the stack. This causes the new state to be pushed down the component tree.

It is currently the responsibility of the `AnalysisItemComponent` to decide which component to render for a given `Item` (it contains various down-casting / reflection like logic for deducing the type of `Item` ). In the template, `analysis-item.component.html`, there is a `ngSwitch` statement which switches on the `shape` property of a `KnownShapedFeature` and there are cases for each known type which decide the component to render. 

Here's a brief overview of the steps required to add a new display shape, say Segments (we still need a couple more shapes to fully support all the current Vamp plugins... i think Segments and Regions / Bars):

- Add a union case to `HigherLevelFeatureShape`, `'segments'`
- Add a union case to `ShapedFeatureData`, an appropriate data type
- Create a type called `Segments`, probably by using `ShapedFeature` - e.g. `export class Segments extends ShapedFeature<'segments', WhateverTypeUsedAbove> {}`
- Add a union case to `KnownShapedFeature` for the above type
- Add to `deduceHigherLevelFeatureShape` to return `'segments'` when the response looks a certain way
- Add to `toKnownShape` to reshape the response data to the segment-like data type when the deduced shape is `'segments'`, and return something conforming to`Segments`.
- Add a `*ngSwitchCase="'segments'"` to an appropriate component inside the `ngSwitch` on the shape in `AnalysisItemComponent` template file.

This will likely involve creating a new angular component which takes the segment-like data type as an `@Input`, because we rely heavily on waves-ui most of the current components extend `WavesComponent` in some way.

So, currently changes have to be made in two files - with most of the changes happening in `FeatureUtilities`. There is room to refactor this, I can imagine a way of "registering" components and known features and setting the app up in such a way in the bootstrapping so that these things fall out dynamically and cases don't have to be added to the `ngSwitch` in the template, but it might not be worth it. 

Lots of parts, and the TypeScript compiler doesn't help you like most functional languages would do regarding a missing case for a discriminated union. Because none of this is actually real, I often wonder  if it might be better to adopt a more object orientated approach, where a lot of these switch statements might disappear via dispatching to a method of some actual concrete type implementing a currently non-existent interface.

**Comment by cannam** [comment]
All of the stuff discussed above is in ugly-duckling, right?

Then we have the feature reshaping code in piper-js (in `HigherLevelUtilities`). This corresponds more-or-less to our conception of the basic higher-level API for Piper, and is very close to what is implemented in the Python Vamp host API as well. This has fewer feature shapes than the ugly-duckling stuff, being limited to the structural shapes `matrix`, `vector`, `tracks`, and `list`. The ugly-duckling code then (in `deduceHigherLevelFeatureShape`) explodes `list` into either `instants` or `notes` (does this mean it currently just throws away any `list` data that is not one of these?)

NB there are more notes that could contribute to this documentation in issues #6 (about piper-js mostly) and #8 (about both piper-js and ugly-duckling).

**Comment by cannam** [comment]
Another documentation question (though not really about ugly-duckling):

* Say I want to write an HTML page and embed some JS in it to load an audio file (from the same server as the page) and extract some features. I have waves-ui included as a bundle in the page and I'm going to use that for display. Is there a bundled version of piper-js I could use? If not, how would I make one? If there was, would it be sufficient for this?

The main thing here is, I want the result to be just an HTML page and a couple of JS bundle files - no server other than plain Apache or whatever, and no build process (no webpack etc, at least not after the redistributable bundle has been made).

**Comment by LucasThompson** [comment]
In Ugly Duckling, the Piper client is a hidden implementation detail of the `FeatureExtractionService`. 

Features are emitted as a `KnownShapedFeature`, or an error is thrown (i.e. no shape could be deduced, what we currently see for something like vamp-example-plugins:fixedtempo:tempo).

**Comment by LucasThompson** [comment]
I used to have a build, but I removed a bunch of code relating to that a few months ago when I cleaned up the dependencies. I think I was using babel and browserify. 

A typical thing to do is pull in whatever packages with npm, develop in the same way as you do in a node environment and then use something like webpack or browserify to bundle the whole thing into a single js file. 

I can, however, create a build again. Packaging this up is increasingly important, along with documentation. 

But yes, if there was one it should be sufficient.

**Comment by cannam** [comment]
Indeed, if the description of how to include this in a web page begins "pull in whatever packages with npm" then it's not going to work (at least for this use case).

I would be interested to see what a bundle with dependencies looks like. There are 347 subdirs of node_modules in my piper-js working copy -- I assume most of these (e.g. typescript and babel stuff) would just be build dependencies if the bundle is to be used as plain Javascript.

**Comment by LucasThompson** [comment]
The bundlers go through some process of building a dependency tree by walking through import / require statements from an entry point .js file. So you only end up with packages which are imported. A transpiler like babel adds a bunch of its own code, runtime libraries / polyfills etc but the compiler code itself won't slip in there / or anything TypeScript related other than the output from the compiler. Some bundlers also go through additional processes to eliminate code from packages which isn't actually used, often referred to in the scene as "tree shaking".

**Comment by cannam** [comment]
A minor detail not mentioned above: `FeatureUtilities` (in the ugly duckling application) lives in the `app/visualisations` directory. (A note on why it lives there, and not in, say, `app/services` might be illuminating)
